drivers/frequency/adf4368/iio_adf4368.c

```diff
@@ -0,0 +1,1087 @@
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ * @return 	  - The size of the read data in case of success, error code
+ * 		    otherwise.
+ */
+static int adf4368_iio_read_channel_attr(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv)
+{
+
+	struct adf4368_iio_dev *iio_adf4368 = (struct adf4368_iio_dev *)dev;
+	struct adf4368_dev *adf4368;
+	int64_t val = -EINVAL;
+	bool phase_pol = false;
+	int32_t opwr;
+	uint64_t freq;
+	uint32_t phase;
+	bool en;
+	int ret;
+
+	if (!iio_adf4368)
+		return -EINVAL;
+
+	adf4368 = iio_adf4368->adf4368_dev;
+
+	if (!adf4368)
+		return -EINVAL;
+
+	switch (priv) {
+	case ADF4368_IIO_CH_ATTR_ENABLE:
+		ret = adf4368_get_en_chan(adf4368, channel->ch_num, &en);
+		if (ret)
+			return ret;
+
+		val = en;
+		break;
+
+	case ADF4368_IIO_CH_ATTR_FREQ:
+		ret = adf4368_get_rfout(adf4368, &freq);
+		if (ret)
+			return ret;
+
+		val = freq;
+		break;
+	case ADF4368_IIO_CH_ATTR_OPWR:
+		ret = adf4368_get_out_power(adf4368, channel->ch_num, &opwr);
+		if (ret)
+			return ret;
+
+		val = opwr;
+		break;
+	case ADF4368_IIO_CH_ATTR_PHASE:
+		ret = adf4368_get_phase(adf4368, &phase, &phase_pol);
+		if (ret)
+			return ret;
+
+		val = phase_pol ? -phase : phase;
+		break;
+	default:
+		return -EINVAL;
+	};
+
+	return snprintf(buf, len, "%"PRIu64, val);
+}
+
+/**
+ * @brief Handles the write request for the channel attributes.
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ * @return 	  - The size of the read data in case of success, error code
+ * 		    otherwise.
+ */
+static int adf4368_iio_write_channel_attr(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv)
+{
+
+	struct adf4368_iio_dev *iio_adf4368 = (struct adf4368_iio_dev *)dev;
+	struct adf4368_dev *adf4368;
+	int64_t val = -EINVAL;
+	bool phase_pol = false;
+
+	int ret;
+
+	if (!iio_adf4368)
+		return -EINVAL;
+
+	adf4368 = iio_adf4368->adf4368_dev;
+
+	if (!adf4368)
+		return -EINVAL;
+
+	sscanf(buf, "%"PRIu64, &val);
+
+	switch (priv) {
+	case ADF4368_IIO_CH_ATTR_ENABLE:
+		ret = adf4368_set_en_chan(adf4368, channel->ch_num, val);
+		break;
+
+	case ADF4368_IIO_CH_ATTR_FREQ:
+		ret = adf4368_set_rfout(adf4368, val);
+		break;
+
+	case ADF4368_IIO_CH_ATTR_OPWR:
+		ret = adf4368_set_out_power(adf4368, channel->ch_num, val);
+		break;
+
+	case ADF4368_IIO_CH_ATTR_PHASE:
+		phase_pol = val < 0 ? true : false;
+		val = abs(val);
+		ret = adf4368_set_phase(adf4368, val, phase_pol);
+		break;
+	default:
+		return -EINVAL;
+	};
+
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static struct iio_attribute adf4368_iio_attrs[] = {
+	{
```

Comment:
could you please to the same for device attributes like you did for channel attributes ? 2 functions that manage all these device attributes, it will result in much more compact code.

adf4368_iio_read_device_attr()
adf4368_iio_write_device_attr()

sorry i didn't see this last pass...

Reply:
and just for completeness, it's not even mandatory to use single function for all, but they just seem so similar that in this case i think it justifies a single function

you could even have a scenario where 1 attribute is so different than the others that you may wanna process it in its own function and have the others be treated in the main function with the big switch case, that's fine too...

Reply:
or I can make separated snprintf for every case

Reply:
Fixed